feat(authz): roles

itayd opened the pull request:
each org member has roles. if "admin" in roles, it is an org admin and has extra privilidges such as inviting and remove users, updating their roles, etc.

itayd commented on tests/system/testdata/orgs/roles.txtar:
```diff
@@ -0,0 +1,79 @@
+
+
+-- test-config.yaml --
+exclusive: true
```
```suggestion
```

daabr reviewed (commented):
After ignoring proto-gen files and trivial stuff, there are still 26 large diffs to review, so this going to take a while...

efiShtain commented on sdk/sdkclients/sdkorgsclient/client.go:
```diff
@@ -55,6 +53,21 @@ func (c *client) GetByID(ctx context.Context, oid sdktypes.OrgID) (sdktypes.Org,
 }
 
+func (c *client) BatchGetByIDs(ctx context.Context, oids []sdktypes.OrgID) ([]sdktypes.Org, error) {
+	resp, err := c.client.BatchGet(ctx, connect.NewRequest(&orgsv1.BatchGetRequest{
```
we have to have a limit here, since it is a public API

itayd replied:
wdym by limit? max num of entries in batch?

efiShtain replied:
yes, 
we can't let users ask for like 1000 objects

daabr replied:
I would add this limit at the lowest (i.e. DB) layer

itayd replied:
added in db

efiShtain commented on proto/autokitteh/orgs/v1/svc.proto:
```diff
@@ -22,6 +22,14 @@ message GetResponse {
   Org org = 1;
 }
 
+message BatchGetRequest {
```
why batch and not list as the other conventions ?

itayd replied:
because we don't list, we get a specific set of entries by their ids.

efiShtain replied:
semantic but ok

efiShtain commented on proto/autokitteh/orgs/v1/org.proto:
```diff
@@ -13,6 +13,7 @@ message OrgMember {
   string user_id = 1;
   string org_id = 2;
   OrgMemberStatus status = 3;
+  repeated string roles = 4;
```
do we want / need multiple roles ?
is this future proofing stuff?

itayd replied:
yes. this is definitely going to happen. we can discuss if you'd like.

itayd replied:
for example, if we'd like to allow all to invite others, we can have an "inviter" role.

efiShtain replied:
stating it again, 
this is future proofing and unrequired feature

efiShtain commented on internal/kittehs/fs.go:
```diff
@@ -57,6 +57,8 @@ func MapToMemFS(in map[string][]byte) (fs.FS, error) {
 	return memfs, nil
 }
 
+func TxtarStringToFS(s string) (fs.FS, error) { return TxtarToFS(txtar.Parse([]byte(s))) }
```
is this for testing only?
why do we need this wrapper?

itayd replied:
i'll remove.

itayd replied:
(used before, seems i stopped using it)

efiShtain commented on internal/backend/db/dbgorm/scheme/records.go:
```diff
@@ -583,14 +583,37 @@ func ParseOrg(r Org) (sdktypes.Org, error) {
+	OrgID  uuid.UUID      `gorm:"primaryKey;type:uuid;not null"`
+	UserID uuid.UUID      `gorm:"primaryKey;type:uuid;not null"`
+	Status int            `gorm:"index"`
+	Roles  datatypes.JSON // [str]
```
why is this a list ?

daabr replied:
I for one support this. Supporting a list of roles instead of a single one is critical on day 1.

Although eventually they shouldn't be in this table but rather in a separate users-to-roles (many-to-many) mappings table, because we need to be able to mass-remove roles without searching and updating all users.

Also, role names should be modifiable, so again storing roles by name is problematic - but not urgent to fix.

But again, supporting multiple roles is critical.

itayd committed:
work

daabr commented on configs/opa_bundles/default/authz.rego:
```diff
@@ -29,52 +37,44 @@ single_associated_project_id(pid) if {
 # Users
 #
 
+# A user can read its own user object.
```
```suggestion
# A user can read their own user object.
```

itayd replied:
done

efiShtain commented on internal/backend/db/dbgorm/orgs.go:
```diff
@@ -42,6 +46,27 @@ func (gdb *gormdb) GetOrg(ctx context.Context, oid sdktypes.OrgID, n sdktypes.Sy
+		return nil, fmt.Errorf("failed to parse orgs: %w", err)
+	}
+
+	if kittehs.ContainedIn(oids...)(authusers.DefaultOrg.ID()) {
```
if it's contained in oids, why wasn't it fetched from db?

itayd replied:
because maybe the oid isn't there

efiShtain commented on internal/backend/db/dbgorm/orgs.go:
```diff
@@ -107,62 +132,64 @@ func (gdb *gormdb) UpdateOrg(ctx context.Context, o sdktypes.Org, fm *sdktypes.F
-func (gdb *gormdb) AddOrgMember(ctx context.Context, oid sdktypes.OrgID, uid sdktypes.UserID, status sdktypes.OrgMemberStatus) error {
-	if oid == authusers.DefaultOrg.ID() {
+func (gdb *gormdb) AddOrgMember(ctx context.Context, m sdktypes.OrgMember) error {
+	if m.OrgID() == authusers.DefaultOrg.ID() {
```
not sure why this check is on the db level and not application level?

daabr replied:
@efiShtain the RPC client is unreliable and optional, and the RPC server needs to be as lean as possible (if we split between the server and the DB, which we do). Checking auth on the lowest level actually makes sense to me as a last line of defense.

itayd replied:
in the future i think i'll move default user and orgs to be seeded instead of being special.

efiShtain commented on internal/backend/db/dbgorm/orgs.go:
```diff
@@ -107,62 +132,64 @@ func (gdb *gormdb) UpdateOrg(ctx context.Context, o sdktypes.Org, fm *sdktypes.F
 
-func (gdb *gormdb) UpdateOrgMemberStatus(ctx context.Context, oid sdktypes.OrgID, uid sdktypes.UserID, s sdktypes.OrgMemberStatus) error {
-	if oid == authusers.DefaultOrg.ID() {
+func (gdb *gormdb) UpdateOrgMember(ctx context.Context, m sdktypes.OrgMember, fm *sdktypes.FieldMask) error {
```
why the field mask is propogated to the db layer ? 
it should be at the application layer or maybe even the tranport layer as a validation

itayd replied:
the db needs to know what fields to update. this is a good representation as any?

efiShtain commented on internal/backend/orgsgrpcsvc/svc.go:
```diff
@@ -201,14 +207,21 @@ func (s *server) ListMembers(ctx context.Context, req *connect.Request[orgsv1.Li
 	}
 
+	var pbusers map[string]*usersv1.User
+	if msg.IncludeUsers {
```
can we move this down to the db layer and just do list members with a join to get the users ?

i think we should try to use as little logic as possible in this layer

itayd replied:
ended up getting rid of an includes all together and always returning the users/orgs. moved the logic for fetching to the db, but not using join yet - will do later on. this simplifies a lot of things as we don't need the BatchXXX methods anymore.

efiShtain commented on internal/backend/orgsgrpcsvc/svc.go:
```diff
@@ -201,14 +207,21 @@ func (s *server) ListMembers(ctx context.Context, req *connect.Request[orgsv1.Li
 
+	var pbusers map[string]*usersv1.User
+	if msg.IncludeUsers {
+		users, err := s.users.BatchGetByIDs(ctx, kittehs.Transform(ms, func(o sdktypes.OrgMember) sdktypes.UserID { return o.UserID() }))
```
can you rename IncludeUsers to something else ?
maybe "FullDetails" or something of that sort

The api is called list members, include users sounds like a membercan be something other than user

itayd replied:
i follow the "include" pattern elsewhere... don't see much issues with it.

efiShtain commented on internal/backend/orgsgrpcsvc/svc.go:
```diff
@@ -224,28 +237,26 @@ func (s *server) GetOrgsForUser(ctx context.Context, req *connect.Request[orgsv1
-		pborgs = kittehs.ListToMap(orgs, func(o *sdkservices.OrgWithMemberStatus) (string, *orgsv1.Org) {
-			pborg := o.Org.ToProto()
-			return pborg.OrgId, pborg
+		orgs, err := s.orgs.BatchGetByIDs(ctx, kittehs.Transform(ms, func(o sdktypes.OrgMember) sdktypes.OrgID { return o.OrgID() }))
```
see my comment above, 
if we push this down to the db layer we can reuse the logic and get better performance (neglible, but cleaner)

efiShtain commented on internal/backend/orgsgrpcsvc/svc.go:
```diff
@@ -266,18 +277,12 @@ func (s *server) GetMember(ctx context.Context, req *connect.Request[orgsv1.GetM
-			Status: status.ToProto(),
-		},
-	}), nil
+	return connect.NewResponse(&orgsv1.GetMemberResponse{Member: m.ToProto()}), nil
```
do we even need get member in the API?

itayd replied:
why not?

itayd replied:
i mean the q should be why yes, but sure - if you want to know only this specific member roles instead of listing everybody.

efiShtain commented on internal/backend/orgs/orgs.go:
```diff
@@ -54,7 +56,9 @@ func (o *orgs) Create(ctx context.Context, org sdktypes.Org) (sdktypes.OrgID, er
 		}
 
-		if err = AddMember(ctx, tx, oid, authcontext.GetAuthnUserID(ctx), sdktypes.OrgMemberStatusActive); err != nil {
+		m := sdktypes.NewOrgMember(oid, authcontext.GetAuthnUserID(ctx)).WithStatus(sdktypes.OrgMemberStatusActive).WithRoles(OrgAdminRoleName)
```
NIT, can you make the line shorter?

itayd replied:
done

daabr commented on internal/backend/gormkitteh/z.go:
```diff
@@ -15,7 +15,7 @@ func OpenZ(z *zap.Logger, cfg *Config, f func(*gorm.Config)) (*gorm.DB, error) {
 		l := &zapgorm2.Logger{
 			ZapLogger:                 z,
-			LogLevel:                  logger.Warn,            // prevent logging SQL queries by default
+			LogLevel:                  logger.Info,
```
This should remain `Warn` by default.

We do not want to log each and every DB query unless we're specifically debugging DB operations locally.

It leaks sensitive data, and it's very spammy.

itayd replied:
done

daabr commented on internal/backend/auth/authz/input.go:
```diff
@@ -0,0 +1,128 @@
+		}
+	})
+
+	// TODO: Use https://docs.styra.com/opa/rego-by-example/builtins/regex/globs_match to match permissions.
```
Add a ticket ID

daabr commented on internal/backend/db/dbgorm/users.go:
```diff
@@ -89,3 +93,24 @@ func (gdb *gormdb) UpdateUser(ctx context.Context, u sdktypes.User, fm *sdktypes
+	q := gdb.db.WithContext(ctx).Where("user_id in ?", kittehs.Transform(uids, func(uid sdktypes.UserID) uuid.UUID { return uid.UUIDValue() }))
+
+	var rs []scheme.User
+	err := q.Find(&rs).Error
```
This is where I would expect to see a hard limit on the number of elements that can be requested

itayd replied:
done

itayd committed:
work

daabr commented on tests/system/auth.go:
```diff
@@ -29,11 +29,40 @@ var (
 
+func addUser(u sdktypes.User) string {
+	return fmt.Sprintf(
+		`insert into users(user_id,email,display_name,created_by,default_org_id) values (%q,%q,%q,%q,%q)`,
```
Never ever EVER use `fmt.Sprintf` to construct a DB query. Not even in test code like here. The potential for abuse is too much - even in test code.

The CodeQL failure below is not a false positive. Either use Gorm, or (less desirable, but acceptable in test code) the suggested fix below.

itayd replied:
now using objects to populate the db

daabr reviewed (changes requested):
The CodeQL failure in this PR must be fixed. In many cases security risks are less or not relevant in test code, but this is not one of these cases. This is easy to abuse for SQL injection, potentially in a real DB too, beyond the expected scope of just system tests.

itayd committed:
wheeee

daabr commented on tests/system/server.go:
```diff
@@ -21,12 +24,32 @@ const (
+		t.Logf("seed object: %s", sobj)
+	}
+
+	path := filepath.Join("/tmp" /*t.TempDir()*/, "autokitteh_seed_objects.json")
```
Bring back `t.TempDir()`

itayd committed:
temp

itayd committed:
work

itayd committed:
feat(authz): add roles, simplify structure + some perm changes (#983)

this simplifies the structure of the authz input to opa, as well as
simplifies a bit the policy itself.
also add more tests.

additionally:
- permit any user to get any other user details.
- deny admins removing themselves from orgs.
- allow non-admins users removing themselves from orgs.

based on #975 .

Ref: ENG-1951.

itayd commented:
merged with a based branch.
